fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MvRewritePreprocessor.java

```diff
@@ -100,27 +104,172 @@ public MvRewritePreprocessor(ConnectContext connectContext,
+        }
+    }
+    
+    /**
+     * To avoid MvRewriteProcessor cost too much optimizer time, reduce all related mvs to a limited size.
+     * <h3>Why to Choose The Best Related MVs Strategy</h3>
+     *
+     * <p>Why still to choose limited related mvs from all active mvs?</p>
+     *
+     * <p>1. optimizer time cost. Even there is MVPlanCache to reduce mv optimizer plan time, but it still may cost
+     * too much time for mv preprocessor, because mv optimizer plan costs too much time because of cold start when
+     * MVPlanCache has no cache, or MVPlanCache has exceeded limited capacity which is 1000 by default.</p>
+     *
+     * <p>2. check mv's freshness for each mv will cost too much time if there are too many mvs.</p>
+     *
+     * <h3>How to Choose The Best Related MVs Strategy</h3>
+     * <p>
+     *     Choose the best related mvs from all active mvs as following order:
+     *     1. find the max intersected table num between mv and query which means it's better for rewrite.
+     *     2. find the latest fresh mv which means its freshness is better.
+     * </p>
+     *
+     * <h3>More Information</h3>
+     * <p>
+     *  NOTE: there are still some limitations about this ordering algorithm:
+     *  1. consider repeated tables in one mv later which one table can be used repeatedly in one mv.
+     *  2. consider random factor so can cache and use more mvs.
+     * </p>
+     */
+    static class MVRelatedOrdering implements Comparator<MaterializedView> {
+        private final Set<String> queryTableNames;
+        private final int queryScanOpNum;
+        private final Map<MaterializedView, MVWithPlanContextCache> mvWithPlanContextCacheMap;
+
+        public MVRelatedOrdering(Set<Table> queryTables,
+                                 int queryScanOpNum,
+                                 Map<MaterializedView, MVWithPlanContextCache> mvPlanContextsMap) {
+            this.queryTableNames = queryTables.stream()
+                    .map(t -> t.getName())
+                    .collect(Collectors.toSet());
+            this.queryScanOpNum = queryScanOpNum;
+            this.mvWithPlanContextCacheMap = mvPlanContextsMap;
+        }
+
+        /**
+         * Get mv and a query's interacted table num. Bigger is Better.
+         */
+        private long getInteractedTablesNum(MaterializedView mv) {
+            List<BaseTableInfo> baseTableInfos = mv.getBaseTableInfos();
+            return baseTableInfos.stream()
+                    .filter(baseTableInfo -> {
+                        String baseTableName = baseTableInfo.getTableName();
+                        if (Strings.isNullOrEmpty(baseTableName)) {
+                            return false;
+                        }
+                        return queryTableNames.contains(baseTableName);
+                    }).count();
+        }
+
+
+        /**
+         * Get mv's base tables' number and query scan op number's diff number. Bigger is Better.
```

Comment:
Does it make more sense to compare the intersected table number instead of NumBaseTable - NumScan ?

Reply:
1. intersected table numbers is better if it's larger.
2. but there are some cases that it's bad if we only considered instersected tables:
case1: If mv contains multi same tables
case2: If mv contains extra tables(with uk/fk constraints)

both in case1 and case2, we prefer a mv that has closest scan ops with query.